src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests.cs

```diff
@@ -1793,5 +1758,31 @@ class C { }
             Assert.Single(classes);
             Assert.Equal("E", classes[0].Identifier.ValueText);
         }
+
+        [Fact]
+        public void Incremental_Generators_Can_Be_Cancelled()
+        {
+            var source = @"
+class C { }
+";
+            var parseOptions = TestOptions.RegularPreview;
+            Compilation compilation = CreateCompilation(source, options: TestOptions.DebugDll, parseOptions: parseOptions);
+            compilation.VerifyDiagnostics();
+
+            Assert.Single(compilation.SyntaxTrees);
+
+            CancellationTokenSource cts = new CancellationTokenSource();
+            var generator = new IncrementalGeneratorWrapper(new PipelineCallbackGenerator((ctx) =>
+            {
+
+                var step1 = ctx.CompilationProvider.Select((c, ct) => { cts.Cancel(); return c; });
+                var step2 = step1.Select((c, ct) => { ct.ThrowIfCancellationRequested(); return c; });
+
+                ctx.RegisterSourceOutput(step2, (spc, c) => spc.AddSource("a", ""));
+            }));
+
+            GeneratorDriver driver = CSharpGeneratorDriver.Create(new ISourceGenerator[] { generator }, parseOptions: parseOptions);
+            Assert.Throws<OperationCanceledException>(() => driver = driver.RunGenerators(compilation, cancellationToken: cts.Token));
```

Comment:
This test doesn't make it clear whose responsibility it is to throw in this case: the generator or the driver. Think you should set a bool local in the `step1.Select` call so we can later assert that it actually ran and was the source of the throw.